### ISSUE TITLE
Enable -Wconversion compiler flag (was mantis id 134)

### ISSUE THREAD

**Issue by nfeske** [genodelabs/genode#23]
GCC manpage about the Wconversion option:

Warn for implicit conversions that may alter a value. This includes conversions between real and integer, like `abs (x)' when`x' is `double'; conversions between signed and unsigned, like`unsigned ui = -1'; and conversions to smaller types, like `sqrtf (M_PI)'. Do not warn for explicit casts like`abs ((int) x)' and `ui = (unsigned) -1', or if the value is not changed by the conversion like in`abs (2.0)'. Warnings about conversions between signed and unsigned integers can be disabled by using `-Wno-sign-conversion'.

**Comment by m-stein** [comment]
@nfeske: I would start adapting the base repos according to this flag if you think it's still a good idea!?

**Comment by m-stein** [comment]
Of course, as less prioritized work while I'm not able to proceed with other stuff.

**Comment by m-stein** [comment]
My plan would be to push little chunks of adaption (for instance one commit for util/string.cc) while adding the flag only in my tools.conf until I reach the point when I can add the flag to the default opts without doing any harm. This way, my work would be continuously merged without a lot of re-basing effort.

**Comment by nfeske** [comment]
Yes, that is the perfect approach, @m-stein. Thanks for taking on this issue!

Let me share the experience of my recent work on `-Wsuggest-override`, which was a very similar undertaking.

1. Just like you suggested, I enabled the warning (in base/mk/global.mk) on my local topic branch.

2. I concentrated on one architecture (x86_64) on the generic code first. In the build.conf of my x86_64 build directory, I enabled only the base and os repositories.

3. I repeatedly built the whole tree by issuing `make` without arguments, fixed a warning, and built again. To speed up the repetitive process, the `make again` feature of the build system becomes handy. It omits the dependency-generation stage and executed the build step immediately. I usually pipe the build output into less to make it easier searchable (`make 2>&1 | less`). So you can see the next error quickly w/o scrolling. I found the use of parallel builds more of a hindrance than the benefit because it does - of course - not always show you the result of the compilation unit one just attempted to fix. When using -j4 one sees four errors at once (which is nice) but one has to keep them and their fixes in short-term memory all at once. Since this is quite taxing, it's probably best to disable parallel builds.

4. After the generic parts of the os and base built completely, I focused on the various kernels. For each of the 8 kernels, I executed `make run/log`. This way, core and the base framework is built for each kernel.

5. Now I successively enabled the other repositories in the build.conf, in the order of demo, libports, gems, ports, and the dde_* repositories. Thanks to the `make again` trick, this is actually doable on a local machine. Enabling the ports repository is the biggest step (as Qt and VirtualBox reside there). Fortunately, it mostly does not use the default strict warning level (because of the third-party code), except for noux. So there are not too many things to touch. One just have to be patient with building everything once.

6. Next, one needs to build for the different architectures (but not for the various boards at this time). E.g., I did the whole build for x86_32, pbxa9 (as a representative for armv7), rpi (for armv6), riscv (not much to build here).

7. At this point we still haven't covered any SoC-specific drivers. Among the drivers, the sd-card drivers supports the most variety of platform. To I made it my goal to build this driver for all boards that the driver supports. Here, one encounters automatically issues like the platform-session interfaces.

8. Look at the differentiation of specs in base-hw. E.g., build once for trustzone and once without trustzone support.

9. Now we most of the warnings fixes, I added the fixes as a commit to staging (this was commit b3727a9b46dbaaadc3845e2d431ea649a253060d).

9. Finally, to catch all the remaining problems (expected mostly in drivers), I enabled the warning on the staging branch and let the nightly results guide the remaining work. (the subsequent commit d9feaa1875c80914a602620bb7096156689b496e) All the remaining issues went as fixup commits to the first commit to the staging branch. In this phase, the task must become a high priority because it blocks others from using the (now temporarily broken) staging branch.

I also recommend using the `merge_to_staging` branch for a while. But don't let the results fool you. It is good as a cheap double check but it does not cover that much of the code base.

It is a very repetitive and labor-intensive work - definitely best pursued as a background task or as a form of meditation. ;-)

**Comment by m-stein** [comment]
Thanks for this detailed sharing of your experience! I think I could use some meditation once in a while :)

**Comment by nfeske** [comment]
I merged the commit series to staging now. Broken builds are expected at this point.

**Comment by skalk** [comment]
@nfeske please consider fixup commits f4863c7628d191d00a624815774360daff2487b3 and b9dc35a185f341275fe3e01617027df8c5aa1ec4 for staging.

**Comment by nfeske** [comment]
When I adjusted the platform_drv/x86 code, I was already a bit anxious about possibly introducing a regression. Well, as evidenced by our nightly tests, this is what just happened. E.g., libc_vfs_fs_fat fails now on x86_64/qemu/nova.

To take this problem from the critical path, I reverted my platform-driver changes from staging (in https://github.com/nfeske/genode/commit/d4adb8e1796d18c2deba82ddcf9fcf3068232f41) and just disabled the `-Wconversion` warnings for this component.

For reference, my intended changes can be found in https://github.com/nfeske/genode/commit/d1b7c3205bc07a88317785960a34be0e70a49ecb - should anyone be willing to lend a sharp eye spotting my mistake. If not, that's fine by me - as we will replace the driver in the not-too-distant future anyway.
